### ISSUE TITLE
RFE: simplify and make less confusing the manual scanner 'System name' and 'Custom System name'

### ISSUE THREAD

**Issue by i30817** [libretro/RetroArch#16031]
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

This is not a bug, but very bad usability and a lost opportunity.

The manual scanner has a very strange way to specify custom playlists filenames. It _looks_ as if it was intended to allow a custom name and to still direct playlist entries to a system databases
 \ thumbnails in the server but it does not because of the strange way it's organized. It's actually very easy for the opposite to happen, a custom playlist name\filename and no thumbnails at all, and the way to avoid that is... tricky.

Edit: this is made even worse because there are two ways for this to happen, and one of them is the default ('System Name' being '\<Directory Content\>' without that directory being named one of RetroArch platform names).

### Expected behavior

'System name' would be obligatory and _forced_ to one of the systems that RetroArch has databases for. This would correspond to the db_name on playlist entries that is used in the process to find\download thumbs and game info.

The 'Custom System Name' if set would be a filename override for the playlist, which would change the name that appears in the tab\list of the RetroArch gui drivers. 

Custom names wouldn't have a specific system icon (because RetroArch doesn't want to be forced to parse playlists to find them), but that is nothing new, and would have another solution anyway, either a new file format, a prefix for the name chosen by the user or whatever. 

I expected 'Custom' (edit: and '\<Content Directory\>') in System Name not to exist. Or if you want to keep the ability to use custom thumbnail dirs and custom databases, to not be _necessary_ to use 'Custom System Name', and 'Custom' be a flag 'also use the custom system name if any for db_name even at the cost of not using the normal databases or thumbnails', ie it would force setting a custom system name.

 'Custom System Name', would always be available, and if filled in, just replace the filename of the playlist. If not filled in, use the 'System Name' as playlist filename.

I'd also recommend renaming both options, to for example 'System name' and 'Playlist name', and the system name option 'Custom' to 'Use playlist name (mandatory to fill in)', if you'd like to keep that. That's *much* clearer than the strange thing that is there now where it seems useful, but it's just a trap that prevents thumbnails working.

### Actual behavior

To chose a custom system name you have to choose 'Custom' in the 'System Name', losing the ability to specify the actual system in the db_name (it uses the custom system name\final playlist filename), which corresponds to the system in the thumbnails directory, and the database for game information. In short, you lose thumbnails and game info.

While the ability to force a custom name is maybe valuable (to have completely separate directories for thumbnails or a completely seperate database) and is maybe valuable for testing new systems, it's not useful for users and just causes confusion of 'why don't thumbnails download' (because the dir chosen to put them in doesn't exist in the server).

Currently if you want thumbnails and a different name for the playlist, you either use a custom thumbnail downloader like https://github.com/i30817/libretrofuzz (that can download from a server dir to a different named local dir) or scan a playlist in one of the current systems (with all of the obvious problems if that playlist already exists), then rename it. Rescanning if you do that is problematic too because I think the system name is picked up from the filename because there is no single equivalent for the multiple db_name properties (empty playlists are legal too).

Fixing this might have the consequence of requiring a new playlist top level scan_* property to record the chosen 'System Name' (since it would no longer be the same as the filename). Note not the proposed 'Playlist Name' because the user should be able to rename the playlist and everything keep working, that one actually would use the filename.

Those scan entries are important to fill in the values on rescan of manual playlists. It's ironic because it would only be used for individual all the same db_names, but I understand why that was done, polysystem playlists would be useful things and there already is a example of one too (the history list). Just not in this case. 

I don't get why the .lpl extension was added though, if all the uses for that string chop it off.

**Comment by i30817** [comment]
@jdgleaver @sonninnos

**Comment by i30817** [comment]
Somehow, when opening this issue, I forgot to check the manual scanner. And the situation is even more fragile than I expected, because there is a 'default directory' system name default that can _very_ easily break thumbnails.

I'd recommend removing that option and _force_ the user to chose one of the presets before scanning, it's a very strange trap for the unwary user for a very minimal amount of convinience for 'experts'.

**Comment by zoltanvb** [comment]
The uses of the fields branch out to various parts of the codebase, so I am wary to do big changes, but I have done some quick tests, with current nightly version:
- if database is selected, custom system name does not matter (as it says in the sublabel) so there is no quick way to do a manual scan against a known system and have it saved in a custom playlist
- if there is no database, thumbnail match can still happen if the respective thumbnail directory is in place (but it needs to have the exact same name as the custom system)

For the second reason, I think forcing a database is counterproductive. Currently ~50 of the ~260 core info files have no database, which is a significant minority. I think there is no better default for "System name" than content directory - moving it to any existing database would be a wild guess, and having it as Custom would require typing from the user by default.

There are 2 improvements that are easily doable and may help the situation:
- hide Custom system name if system name is not set as custom
- use dynamic sublabel for the actual Scan entry to warn the user if the system name (selected, parsed from dir, or custom) does not match any database name and consequently would not allow any thumbnail download

In addition, a custom playlist name would come in handy as well.

**Comment by i30817** [comment]
> if database is selected, custom system name does not matter (as it says in the sublabel) so there is no quick way to do a manual scan against a known system and have it saved in a custom playlist

Yes. And that is the usability lost opportunity I mention. Besides these being unrelated concerns, besides the playlist entries already specifying the database when they are needed, a custom subset of a single database is actually quite valuable?

For a simple example, scummvm supports thousands of games, a vast majority of them in certain engines only. I personally seperate the AGS engine games from the rest, but I can easily see someone separating the DIRECTOR games too, or ... That other big number of games engine I'm forgetting.

It's not just scummvm either, MAME is another obvious example of why you might not want everything in the same playlist and still get thumbnails\game info.

> For the second reason, I think forcing a database is counterproductive. Currently ~50 of the ~260 core info files have no database, which is a significant minority. I think there is no better default for "System name" than content directory - moving it to any existing database would be a wild guess, and having it as Custom would require typing from the user by default.

While I would prefer content directory didn't exist, I simply think it can still be a option, just not allow the scan to progress until 'a' option is chosen and it starting undefined so a user is forced to choose, instead of the (really really common in the subreddit) situation where the user leaves it with the default value because they don't understand what it does, then gets no thumbnails because the directory they scanned has a random name. No default, obligatory choice and a clean label warning about the content directory option. And getting rid of custom to simplify.

Then renaming 'custom system name' to 'playlist name' and change its function to _just_ change the playlist final name if filled in.

**Comment by i30817** [comment]
If for some reason custom is not removed (for full clarity I can only think of one reason, a console setup where you can't easily rename directories without getting to pc), a still superior alternative is to make 'Custom' a chosen option... 

But don't overload the 'custom system name', ie seperate the database\thumbnail name choice entry from the playlist name, precisely to allow custom subsets. Make custom unhide a custom database\thumbnail name entry.

But id really prefer to eliminate it, because the custom interface is ... not good.

**Comment by i30817** [comment]
As for order, id suggest something like


├── Playlist Name
├── System Name
│   ├── Sony PlayStation (and every other system)
│   ├── <Custom> (if wanted)
│   └── <Content Directory> (if wanted)
└── Custom System Name ( hidden until custom is chosen)

Just so people know that 'Playlist Name' has nothing to do with the system name part.
